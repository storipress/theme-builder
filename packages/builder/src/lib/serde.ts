import { parse, stringify } from 'devalue'
import { isReactive, isRef, isShallow, reactive, ref, shallowRef, toRaw } from 'vue'

const serializeReducers = {
  Reactive: (obj: unknown) => {
    if (isReactive(obj)) {
      // vue 2 is same reference
      return structuredClone(obj)
    }
  },
  Ref: (obj: unknown) => {
    if (isRef(obj) && !isShallow(obj)) {
      return toRaw(obj.value)
    }
  },
  ShallowRef: (obj: unknown) => {
    if (isRef(obj) && isShallow(obj)) {
      return toRaw(obj.value)
    }
  },

  FreezedArray: (obj: unknown) => {
    if (typeof obj === 'object' && Object.isFrozen(obj) && Array.isArray(obj)) {
      return [...obj]
    }
  },
  FreezedObject: (obj: unknown) => {
    if (typeof obj === 'object' && Object.isFrozen(obj) && !Array.isArray(obj) && obj !== null) {
      return {
        ...obj,
      }
    }
  },
}

export function serialize(obj: unknown) {
  return stringify(obj, serializeReducers)
}

const deserializeReducers = {
  Reactive: (obj: unknown) => {
    return reactive(obj as Record<string, unknown>)
  },
  Ref: (obj: unknown) => {
    return ref(obj)
  },
  ShallowRef: (obj: unknown) => {
    return shallowRef(obj)
  },
  FreezedArray: (obj: unknown) => {
    return Object.freeze(obj)
  },
  FreezedObject: (obj: unknown) => {
    return Object.freeze(obj)
  },
}

export function deserialize<T>(json: string): T {
  return parse(json, deserializeReducers)
}
