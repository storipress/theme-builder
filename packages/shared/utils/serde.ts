import { identity, isObject } from 'lodash'

interface SpecialObject {
  $$type: string
  value: unknown
}

function isSpecialObject(x: unknown): x is SpecialObject {
  if (isObject(x) && '$$type' in x) {
    return true
  }
  return false
}

export function serialize(obj: unknown) {
  return JSON.stringify(obj, (_, value) => {
    if (value instanceof Map) {
      return {
        $$type: 'map',
        value: [...value.entries()],
      }
    }
    if (value instanceof Set) {
      return {
        $$type: 'set',
        value: [...value.values()],
      }
    }
    if (Array.isArray(value) && Object.isFrozen(value)) {
      return {
        $$type: 'freezeArray',
        value: [...value],
      }
    }
    if (isObject(value) && Object.isFrozen(value)) {
      return {
        $$type: 'freeze',
        value: { ...value },
      }
    }
    return value
  })
}

function deserializeAs(Ctor: new (x: any) => unknown): (x: unknown) => unknown {
  return (x) => new Ctor(x)
}

const deserializeSpecialObject: Record<string, (x: unknown) => unknown> = {
  map: deserializeAs(Map),
  set: deserializeAs(Set),
  freeze: Object.freeze,
  freezeArray: Object.freeze,
}

export function deserialize<T>(json: string): T {
  return JSON.parse(json, (_, value) => {
    if (isSpecialObject(value)) {
      const des = deserializeSpecialObject[value.$$type] ?? identity
      return des(value.value)
    }
    return value
  })
}
