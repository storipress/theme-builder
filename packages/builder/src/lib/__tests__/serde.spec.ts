import { expect, it } from 'vitest'

import { deserialize, serialize } from '../serde'

it('handle freezed object', () => {
  const serialized = serialize(Object.freeze({ a: { b: 1 } }))
  expect(serialized).toMatchSnapshot()
  const obj: { a: { b: number } } = deserialize(serialized)
  expect(obj).toEqual({ a: { b: 1 } })
  expect(Object.isFrozen(obj)).toBe(true)
  expect(Object.isFrozen(obj.a)).toBe(false)
})

it('handle freezed array', () => {
  const serialized = serialize(Object.freeze([[1], [2]]))
  expect(serialized).toMatchSnapshot()
  const obj: number[][] = deserialize(serialized)
  expect(obj).toEqual([[1], [2]])
  expect(Object.isFrozen(obj)).toBe(true)
  expect(Object.isFrozen(obj[0])).toBe(false)
})

it('handle Map', () => {
  const serialized = serialize(new Map([['a', 1]]))
  const obj: Map<string, number> = deserialize(serialized)
  expect(obj).toEqual(new Map([['a', 1]]))
})

it('handle Set', () => {
  const serialized = serialize(new Set([1, 2]))
  const obj: Set<number> = deserialize(serialized)
  expect(obj).toEqual(new Set([1, 2]))
})
