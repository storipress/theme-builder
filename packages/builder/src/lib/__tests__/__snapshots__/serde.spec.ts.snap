// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handle freezed array 1`] = `"[["FreezedArray",1],[2,4],[3],1,[5],2]"`;

exports[`handle freezed object 1`] = `"[["FreezedObject",1],{"a":2},{"b":3},1]"`;
